### PR TITLE
ZEN-24152 Move productionState into a separate lookup table

### PR DESCRIPTION
**Pull request by johnlhamilton**
This change replaces the productionState and preMWProductionState attributes on ManagedEntity with getter and setter methods that store this information in a separate lookup table.  We also removed production state from the catalogs and handle searching and sorting on production state by looking up the values in the lookup table.

IMPORTANT:
1. This includes a migration that must be run during upgrade.
2. This change requires an update to Acquisition, using our forked version here:  https://github.com/zenoss/Acquisition/pull/1

### CODE REVIEW

File: Products/ZenUtils/productionstate/productionstate.py [hunk 1]
```diff
@@ -16,35 +16,62 @@
 from zope.interface import implements
 from .interfaces import IProdStateManager
 
+from BTrees.OOBTree import OOBTree
+from Products.ZenUtils.guid.interfaces import IGlobalIdentifier
+
+PRODSTATE_TABLE_PATH = '/zport/dmd/prodstate_table'
 DEFAULT_PRODUCTION_STATE = 1000
 
+class ProdState(object):
```
Comment (johnlhamilton) [inline Products/ZenUtils/productionstate/productionstate.py]: Should this be Persistent?  Or, just use a tuple.

File: Products/ZenUtils/productionstate/productionstate.py [hunk 2]
```diff
@@ -16,35 +16,62 @@
...
-            return DEFAULT_PRODUCTION_STATE
-
-        return object._preMWProductionState
+        guid = IGlobalIdentifier(object).getGUID()
+        pstate = self._getProdStatesFromTable(guid).preMWProductionState
+        if pstate is None:
+            return self._default_state
+        return pstate
 
     def setProductionState(self, object, value):
```
Comment (johnlhamilton) [inline Products/ZenUtils/productionstate/productionstate.py]: use obj, not object

File: Products/ZenWidgets/browser/Portlets.py [hunk 3]
```diff
@@ -91,16 +91,16 @@ def getDevProdStateJSON(self, prodStates=['Maintenance']):
...
-        for state in prodStates:
-            queries.append(Eq('getProdState', state))
-        query = Or(*queries)
+        query = Indexed('id') # Get all the brains
         objects = catalog.evalAdvancedQuery(query, ((orderby, orderdir),))
+
         devs = (x.getObject() for x in objects)
         mydict = {'columns':['Device', 'Prod State'], 'data':[]}
         for dev in devs:
+            if not dev.getProductionStateString() in prodStates: continue
```
Comment (johnlhamilton) [inline Products/ZenWidgets/browser/Portlets.py]: This is too slow, use the global catalog so we don't have to wake up every device

File: Products/Zuul/facades/__init__.py [hunk 4]
```diff
@@ -181,17 +184,87 @@ def getDeviceBrains(self, uid=None, start=0, limit=50, sort='name',
...
+                productionStates.sort(reverse=reverse)
+                prodStateBuckets = {}
+                for ps in productionStates:
+                    prodStateBuckets[ps] = []
+
+                for brain in psFilteredbrains:
+                    prodState = psManager.getProductionStateFromGUID(brain.uuid)
+                    prodStateBuckets[prodState].append(brain)
+
+                sortedBrains = list(brain for brain in mergeBuckets(productionStates, prodStateBuckets))
```
Comment (johnlhamilton) [inline Products/Zuul/facades/__init__.py]: drop 'list'

**Comment by johnlhamilton** [comment]
Need to include removal of indexes in the migration

File: Products/ZenModel/ManagedEntity.py [hunk 5]
```diff
@@ -68,17 +69,63 @@ def device(self):
...
     def setPreMWProductionState(self, state):
-        IProdStateManager(self.dmd).setPreMWProductionState(self, state)
+        self.getProdStateManager().setPreMWProductionState(self, state)
+
+    # In order to maintain backward-compatibility, we need to preserve productionState as a property.
+    #  Our getProductionState() method requires acquisition context, so we have to add the property
+    #  onto the wrapped object and not on the ManagedEntity object itself.  We do this by sub-classing 
+    #  Zope's wrapper. Note that sub-classing sope's wrapper only works because we modified the Acquisition
+    #  source code to handle it properly. 
+    def __of__(self, ob):
```
Comment (johnlhamilton) [inline Products/ZenModel/ManagedEntity.py]: lets call this container

File: Products/ZenModel/ManagedEntity.py [hunk 6]
```diff
@@ -73,54 +74,68 @@ def getProdStateManager(self):
...
         self.getProdStateManager().setPreMWProductionState(self, state)
 
+    def resetProductionState(self):
+        self.getProdStateManager().clearProductionState(self)
+
     # In order to maintain backward-compatibility, we need to preserve productionState as a property.
     #  Our getProductionState() method requires acquisition context, so we have to add the property
     #  onto the wrapped object and not on the ManagedEntity object itself.  We do this by sub-classing 
     #  Zope's wrapper. Note that sub-classing sope's wrapper only works because we modified the Acquisition
     #  source code to handle it properly.
```
Comment (johnlhamilton) [inline Products/ZenModel/ManagedEntity.py]: add link to acquisition pr here

File: Products/ZenUtils/productionstate/interfaces.py [hunk 7]
```diff
@@ -11,6 +11,10 @@
 from exceptions import Exception
 from zope.interface import Interface
 
+class ProdStateNotSetError(Exception):
+    def __init__(self,*args,**kwargs):
```
Comment (johnlhamilton) [inline Products/ZenUtils/productionstate/interfaces.py]: just pass/docstring

File: Products/ZenModel/migrate/moveProductionStateToBTree.py [hunk 8]
```diff
@@ -23,19 +25,33 @@ class MoveProductionStateToBTree(Migrate.Step):
...
             obj.setPreMWProductionState(obj_unwrapped.preMWProductionState)
             del obj_unwrapped.preMWProductionState
+            migrated=True
+        return migrated
 
     def cutover(self, dmd):
+        log.info("Migrating productionState to lookup table")
+        total=len(dmd.Devices.getSubDevices_recursive())
+        count = 1
         for device in dmd.Devices.getSubDevices_recursive():
```
Comment (johnlhamilton) [inline Products/ZenModel/migrate/moveProductionStateToBTree.py]: drop _recursive

File: Products/ZenModel/migrate/moveProductionStateToBTree.py [hunk 9]
```diff
@@ -23,19 +25,33 @@ class MoveProductionStateToBTree(Migrate.Step):
...
-            self.migrateObject(device)
+            log.info("Checking if productionState migration required for device %d of %d (%s)", count, total, device)
```
Comment (johnlhamilton) [inline Products/ZenModel/migrate/moveProductionStateToBTree.py]: too chatty, update less frequently

File: Products/ZenModel/migrate/moveProductionStateToBTree.py [hunk 10]
```diff
@@ -23,19 +25,33 @@ class MoveProductionStateToBTree(Migrate.Step):
...
+            migrated = self.migrateObject(device)
 
             # migrate components
             for c in device.getDeviceComponents():
-                self.migrateObject(c)
+                cMigrated = self.migrateObject(c)
+                migrated = migrated or cMigrated
+
+            if migrated:
+                log.info("Successfully migrated productionState for %s", device)
```
Comment (johnlhamilton) [inline Products/ZenModel/migrate/moveProductionStateToBTree.py]: too chatty, remove

File: Products/ZenWidgets/browser/Portlets.py [hunk 11]
```diff
@@ -88,19 +89,26 @@ def getDevProdStateJSON(self, prodStates=['Maintenance']):
...
+            for t in self.context.getProdStateConversions():
+                if t[0] == prodStateString:
+                    return t[1]
+
+        numericProdStates = [getProdStateInt(p) for p in prodStates]
+
         orderby, orderdir = 'id', 'asc'
-        catalog = getattr(devroot, devroot.default_catalog)
-        query = Indexed('id') # Get all the brains
+        catalog = self.context.getPhysicalRoot().zport.global_catalog
```
Comment (fraibacas) [inline Products/ZenWidgets/browser/Portlets.py]: Other way to get the global catalog:
from Products.Zuul.catalog.interfaces import IModelCatalogTool
global_catalog = IModelCatalogTool(self.context)

**Comment by iancmcc** [comment]
retest this

**Comment by iancmcc** [comment]
This passed when we rekicked the job manually in jenkins after some changes to platform-build. The job wasn't updated, but for the sake of time I'm merging this.

http://jenkins.zendev.org/job/zenoss-prodbin-pull-requests-develop/712 is the successful PR build.

**Pull request merged**